### DISCUSSION TITLE
Traffic handling across Tabs in Cypress browser

### DISCUSSION THREAD

**Discussion by metalix2** [cypress-io/cypress#28337] [General]
Hello, 

I wanted to raise a discussion to get some insights on how Cypress plans to support network traffic on other tabs.
In previous versions of Cypress, traffic on any tab/page of the browser were handled the same. If you had an intercept for a test it would also be applied to other tabs in the browser.

I understand that the cypress browser loads the suite page as it's main entry point for running tests. However there are other apps which also benefited from running through Tabs such as lighthouse. It was pretty useful having a single mocking solution to run e2e, visual and performance tests. If Cypress has no plans to support other tabs or keep them serperate to the cypress testing suite it would be good to know that sooner rather than later. 

Over that past few versions of Cypress the network behaviour around tabs has changed quite dramatically.
Going from Slow to Slower, missing headers and now behaving as if just a regular Chrome tab.

What is the expected behaviour for a new tab in the Cypress browser?

Looking at recent Pull requrest comments it looks like the aim is to have new tabs behave as normal Tabs. 

> For extra tabs/windows, we skip a lot of middleware for the sake of letting the response go through un-modified.
https://github.com/cypress-io/cypress/pull/28322#issue-1990778198

> Filters out network requests from extra tabs and windows (extra meaning not the main Cypress tab) so that they bypass most of the proxy logic. They aren't correlated, awaited, injected into, etc. The intent is for the requests and responses to go through as unchanged as possible.
https://github.com/cypress-io/cypress/pull/28188#issue-1970514748

**Comment by jennifer-shehane** [comment]
We've been making several changes to the way tabs work within Cypress with the aim of having a beta [`@cypress/puppeteer` plugin](https://github.com/cypress-io/cypress/pull/28370) released to enable some use cases of testing across multiple tabs. 

It was never intended for Cypress to interact in any way with other tabs and we even display a warning when opening new tabs within Cypress discouraging the use of other tabs. 

![Screen Shot 2023-11-21 at 2 26 02 PM](https://github.com/cypress-io/cypress/assets/1271364/cf8c2aac-5e78-4800-adc0-2634805c010d)

With the work around getting the plugin to function and wanting the tabs to not interact any way with Cypress test runs, we've closed some situations where there were interactions between the new that was never intended and actually causing issues in our proxy logic like https://github.com/cypress-io/cypress/pull/28188.

The 13.6.0 release today will see some additional changes to how tabs work in order to further faciliate the plugin to offer some forms of multi-tab testing, like closing all tabs in between each test run. 

13.6.0 contains the final changes we intended to make for the beta release of the plugin, wherein we will gather feedback from the community. 

We'd love to hear your feedback on how these changes affect your use of Cypress so that we can use that feedback to deliver a better product.

**Reply by metalix2** [reply]
Hi Jennifer,

Thank you for your explanation it was exactly what I was looking for. I think this makes sense. Originally when we set up the lighthouse tests with cypress we were surprised that the intercepts did work. Though I can see that it was a bit hacky. 

I believe puppeteer is going to be our best bet going forward being natively supported with lighthouse. Any modifications to the browser should be set up using puppeteer. This[ example](https://github.com/justinribeiro/lighthouse-jest-example) covers it nicely.

I appreciate your time in helping us move forward with this. We'll review the newer versions of Cypress and share our thoughts. One big ask we've seen internally has been the support for http2 But this was related to lighthouse testing. Which may not be relevant any more.

**Comment by jennifer-shehane** [comment]
The [@cypress/puppeteer plugin](https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer) has been released btw, which is what these changes were in preparation for.
